### PR TITLE
add sync_log for kv_rpc_request

### PR DESCRIPTION
**Pull request by javaforfun**
Fixes https://github.com/pingcap/tikv/issues/2271

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
LGTM

File: src/raftstore/store/peer.rs [hunk 1]
```diff
@@ -1086,14 +1089,21 @@ impl Peer {
         self.proposals.push(meta);
     }
 
-    fn get_handle_policy(&mut self, req: &RaftCmdRequest) -> Result<RequestPolicy> {
+    fn get_handle_policy(
+        &mut self,
+        req: &RaftCmdRequest,
+        sync_log: &mut bool,
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: I prefer moving to another function to check whether we need sync or not.

File: tests/raft/test_raft.rs [hunk 2]
```diff
@@ -240,24 +241,27 @@ pub fn new_message(from: u64, to: u64, t: MessageType, n: usize) -> Message {
...
 
-pub fn new_entry(term: u64, index: u64, data: Option<&str>) -> Entry {
+pub fn new_entry(term: u64, index: u64, data: Option<&str>, empty_entry: bool) -> Entry {
     let mut e = Entry::new();
     e.set_index(index);
     e.set_term(term);
     if let Some(d) = data {
         e.set_data(d.as_bytes().to_vec());
     }
+    if !empty_entry {
```
Comment (siddontang) [inline tests/raft/test_raft.rs]: why using empty_entry to determine sync or not?

File: src/raftstore/store/peer.rs [hunk 3]
```diff
@@ -1086,14 +1089,21 @@ impl Peer {
...
+    ) -> Result<RequestPolicy> {
         if req.has_admin_request() {
             if apply::get_change_peer_cmd(req).is_some() {
                 return Ok(RequestPolicy::ProposeConfChange);
             }
             if get_transfer_leader_cmd(req).is_some() {
                 return Ok(RequestPolicy::ProposeTransferLeader);
             }
+            if req.get_admin_request().get_cmd_type() == AdminCmdType::Split {
```
Comment (siddontang) [inline src/raftstore/store/peer.rs]: no need to care ConfChange?

**Comment by siddontang** [comment]
I think we must port https://github.com/coreos/etcd/blob/master/raft/node.go#L530 later.

File: tests/raft/test_raft.rs (hunk 2 above)
Comment (javaforfun) [reply tests/raft/test_raft.rs]: all entries in test_raft are not sync,  but we may have entry.has_sync_log() == false

File: src/raftstore/store/peer.rs (hunk 3 above)
Comment (javaforfun) [reply src/raftstore/store/peer.rs]: it's in propose_conf_change, not propose_normal

File: tests/raft/test_raft.rs (hunk 2 above)
Comment (siddontang) [reply tests/raft/test_raft.rs]: I still can't get the point... 

The default sync_log is false for the event, why do you set to false again with not empty?

**Comment by siddontang** [comment]
LGTM

PTAL @BusyJay

**Comment by siddontang** [comment]
PTAL @BusyJay

File: src/raftstore/store/peer.rs [hunk 4]
```diff
@@ -1695,6 +1698,23 @@ fn get_transfer_leader_cmd(msg: &RaftCmdRequest) -> Option<&TransferLeaderReques
     Some(req.get_transfer_leader())
 }
 
+fn get_sync_log_from_request(msg: &RaftCmdRequest) -> bool {
+    if msg.has_admin_request() {
+        let req = msg.get_admin_request();
+        if req.get_cmd_type() == AdminCmdType::ChangePeer ||
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Why not `return req.get_cmd_type() == AdminCmdType::ChangePeer || req.get_cmd_type() == AdminCmdType::Split`?

File: src/raftstore/store/peer.rs [hunk 5]
```diff
@@ -1695,6 +1698,23 @@ fn get_transfer_leader_cmd(msg: &RaftCmdRequest) -> Option<&TransferLeaderReques
...
+            req.get_cmd_type() == AdminCmdType::Split
+        {
+            return true;
+        }
+        return false;
+    }
+
+    if msg.has_header() && msg.get_header().has_sync_log() && msg.get_header().get_sync_log() {
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: Has check is unnecessary since the default value for `sync_log` is false.

File: tests/raft/test_raft.rs [hunk 6]
```diff
@@ -248,7 +249,10 @@ pub fn new_message(from: u64, to: u64, t: MessageType, n: usize) -> Message {
 }
 
 pub fn empty_entry(term: u64, index: u64) -> Entry {
-    new_entry(term, index, None)
+    let mut e = Entry::new();
```
Comment (BusyJay) [inline tests/raft/test_raft.rs]: Why not use `new_entry`?

File: src/raftstore/store/peer.rs [hunk 7]
```diff
@@ -1695,6 +1698,19 @@ fn get_transfer_leader_cmd(msg: &RaftCmdRequest) -> Option<&TransferLeaderReques
...
 }
 
+fn get_sync_log_from_request(msg: &RaftCmdRequest) -> bool {
+    if msg.has_admin_request() {
+        let req = msg.get_admin_request();
+        return req.get_cmd_type() == AdminCmdType::ChangePeer ||
+            req.get_cmd_type() == AdminCmdType::Split;
+    }
+
+    if msg.has_header() && msg.get_header().get_sync_log() {
```
Comment (BusyJay) [inline src/raftstore/store/peer.rs]: `has_header` is also unnecessary.

**Comment by siddontang** [comment]
Please fix the conflict @javaforfun

**Comment by BusyJay** [comment]
LGTM

**Comment by javaforfun** [comment]
need merge https://github.com/pingcap/kvproto/pull/192  first

**Comment by siddontang** [comment]
@javaforfun 

You can fix the conflict and merge now.

**Pull request merged**